[DirectX] Bug fix for Data Scalarization crash

farzonl opened the pull request:
Two bugs here. First  calling `Inst->getFunction()`  has undefined behavior if the instruction is not tracked to a function. I suspect the `replaceAllUsesWith` was leaving the GEPs in a weird ghost parent situation. I switched up the visitor to be able to `eraseFromParent` as part of visiting and then everything started working.

The second bug was in `DXILFlattenArrays.cpp`. I was unaware that you can have multidimensional arrays of `zeroinitializer`, and `undef` so fixed up the initializer to handle these two cases.

fixes #117273

farzonl committed:
[DirectX] Bug fix for Data Scalarization crash

bogner commented on llvm/lib/Target/DirectX/DXILFlattenArrays.cpp:
```diff
@@ -321,6 +321,18 @@ static void collectElements(Constant *Init,
     Elements.push_back(Init);
     return;
   }
+  auto *LLVMArrayType = dyn_cast<ArrayType>(Init->getType());
```
This should either be `cast<ArrayType>` (since the cast can't fail) or we should do the `dyn_cast` and check for null above instead of `isa`.

This is also kind of a weird name - I think `ArrayTy` is the usual abbreviation.

bogner commented on llvm/lib/Target/DirectX/DXILFlattenArrays.cpp:
```diff
@@ -321,6 +321,18 @@ static void collectElements(Constant *Init,
   }
+  auto *LLVMArrayType = dyn_cast<ArrayType>(Init->getType());
+  if (isa<ConstantAggregateZero>(Init)) {
+    for (unsigned I = 0; I < LLVMArrayType->getNumElements(); ++I)
```
It's best not to call a function in the condition of a for loop. It doesn't really matter much here since this happens to be a simple accessor, but it's good form to avoid it.
```suggestion
    for (unsigned I = 0, E = LLVMArrayType->getNumElements(); I != E; ++I)
```

Same comment a few lines below.

bogner commented on llvm/test/CodeGen/DirectX/flatten-bug-117273.ll:
```diff
@@ -0,0 +1,25 @@
+; RUN: opt -S -passes='dxil-flatten-arrays,dxil-op-lower' -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+
+
+@_ZL7Palette = internal constant [2 x [3 x float]] [[3 x float] zeroinitializer, [3 x float] undef], align 16
```
Might as well change these C++ mangled names to something a bit more human readable - they don't matter to the test.

bogner reviewed (approved):
Just a few style comments. LGTM.

farzonl committed:
1. address PR comments
2. remove use of convertUsersOfConstantsToInstructions (was leaving use
   chains in unstable state).
3. remove RPOT for DXILDataScalarization.cpp. we were visiting twice.

farzonl commented on llvm/lib/Target/DirectX/DXILDataScalarization.cpp:
```diff
@@ -106,6 +90,21 @@ bool DataScalarizerVisitor::visitLoadInst(LoadInst &LI) {
+      GetElementPtrInst *OldGEP =
+          cast<GetElementPtrInst>(CE->getAsInstruction());
+      OldGEP->insertBefore(&LI);
+      Value *NewGEP = createNewGetElementPtr(*OldGEP);
```
Few things going on here.  instead of calling `convertUsersOfConstantsToInstructions` we are going to do something simpler to split the GEP Operand into an instruction.

When you call `getAsInstruction()` off a ConstExpr it creates a new instruction. So that we don't lose track of it and so it gets parented we do `insertBefore`.

`createNewGetElementPtr` wasn't strictly necessary. We could have called `visitGetElementPtrInst` at the end and then we wouldn't need to do `OldGEP->eraseFromParent();` 

Infact calling `visitGetElementPtrInst` is how I'm doing it in `DXILFlattenArrays.cpp` because that pass has to deal with more complicated things to update the GEP chains. Only reason I decided to turn `visitGetElementPtrInst` into a helper is because of how straight forward it was to only operate on one instruction.  I don't have strong feelings on this approach so if you think of something better please let me know.

farzonl commented on llvm/test/CodeGen/DirectX/llc-vector-load-scalarize.ll:
```diff
@@ -33,18 +32,13 @@ define <4 x i32> @load_array_vec_test() #0 {
 ; CHECK-NEXT:    [[TMP7:%.*]] = bitcast ptr addrspace(3) getelementptr (i32, ptr addrspace(3) @arrayofVecData.scalarized.1dim, i32 3) to ptr addrspace(3)
 ; CHECK-NEXT:    [[TMP8:%.*]] = load i32, ptr addrspace(3) [[TMP7]], align 4
-; CHECK-NEXT:    [[TMP9:%.*]] = bitcast ptr addrspace(3) @arrayofVecData.scalarized.1dim to ptr addrspace(3)
-; CHECK-NEXT:    [[TMP10:%.*]] = getelementptr [2 x [3 x float]], ptr addrspace(3) [[TMP9]], i32 0, i32 1
```
This is one of the bugs from `convertUsersOfConstantsToInstructions`. It would split the bitcast from the GEP, but then we would never visit the GEP and so the GEP type never got flattened.

farzonl commented on llvm/test/CodeGen/DirectX/llc-vector-load-scalarize.ll:
```diff
@@ -87,7 +81,7 @@ define <4 x i32> @load_vec_test() #0 {
 ; CHECK-LABEL: define <4 x i32> @load_static_array_of_vec_test(
 ; CHECK-SAME: i32 [[INDEX:%.*]]) #[[ATTR0]] {
-; CHECK-NEXT:    [[DOTFLAT:%.*]] = getelementptr [12 x i32], ptr @staticArrayOfVecData.scalarized.1dim, i32 [[INDEX]]
+; CHECK-NEXT:    [[DOTFLAT:%.*]] = getelementptr inbounds [12 x i32], ptr @staticArrayOfVecData.scalarized.1dim, i32 [[INDEX]]
```
our GEP builders were not preserving inbounds. Now we are.

farzonl commented on llvm/lib/Target/DirectX/DXILFlattenArrays.cpp:
```diff
@@ -317,10 +330,22 @@ bool DXILFlattenArraysVisitor::visit(Function &F) {
+  }
+  if (isa<UndefValue>(Init)) {
+    for (unsigned I = 0; I < ArrSize; ++I)
+      Elements.push_back(UndefValue::get(ArrayTy->getElementType()));
```
Note to myself, we need to figure out what to do with undef for DXIL now that is deprecated.

llvm-beanz commented on llvm/test/CodeGen/DirectX/flatten-bug-117273.ll:
```diff
@@ -0,0 +1,23 @@
+; RUN: opt -S -passes='dxil-flatten-arrays,dxil-op-lower' -mtriple=dxil-pc-shadermodel6.3-library %s | FileCheck %s
+
+
+@ZerroInitAndUndefArr = internal constant [2 x [3 x float]] [[3 x float] zeroinitializer, [3 x float] undef], align 16
```
The test case in #117273 doesn't have anything that should be producing an undef array, and I'm not sure it should be possible to create an undef array in HLSL. Do you know where this is coming from?

farzonl replied:
It wasn't your exact example I just was playing around with how to change the values in  the array after I saw float3(0,0,0) became `zeroinitializer `. I then noticed that if I created an uninitialized array and passed it to the Pallette array I could get an undef array.
 example with fcgl: https://hlsl.godbolt.org/z/oPxxoKf9f
example without: https://hlsl.godbolt.org/z/rr1TocKnj

farzonl committed:
remove nested undef handling

farzonl committed:
address pr comments
